workflow/rules/common.smk

```diff
@@ -28,27 +29,128 @@ validate(samples, schema="../schemas/samples.schema.yaml")
 
 ### Read and validate units file
 
-units = pd.read_table(config["units"], dtype=str).set_index(
-    ["sample", "unit", "run", "lane"], drop=False
-)
+units = pd.read_table(config["units"], dtype=str).set_index(["sample", "unit", "run", "lane"], drop=False)
 validate(units, schema="../schemas/units.schema.yaml")
 
 ### Set wildcard constraints
 
 
 wildcard_constraints:
     sample="|".join(samples.index),
-
-def get_fastq(wildcards):
-    fastqs = units.loc[
-        (wildcards.sample, wildcards.unit, wildcards.run, wildcards.lane), ["fastq1", "fastq2"]
-    ].dropna()
-    return {"fwd": fastqs.fastq1, "rev": fastqs.fastq2}
-
-
-def get_sample_fastq(wildcards):
-    fastqs = units.loc[(wildcards.sample, wildcards.unit), ["fastq1", "fastq2"]].dropna()
-    return {"fwd": fastqs["fastq1"].tolist(), "rev": fastqs["fastq2"].tolist()}
-
-def compile_output_list(wildcards):
-    return ["dummy.txt"]
+    unit="N|T|R",
+    read="fastq[1|2]",
+
+
+def get_unit(units: pd.DataFrame, wildcards: snakemake.io.Wildcards) -> pd.Series:
+    """
+    function used to extract one unit(row) from units.tsv
+
+    Args:
+        units: DataFrame generate by importing a file following schema defintion
+               found in pre-alignment/workflow/schemas/units.schema.tsv
+        wildcards: wildcards object with at least the following widlcard names
+               sample, unit, run, and lane
+
+    Returns:
+        Series containing data of the selected row
+
+    Raises:
+        raises and exception if no unit can be extracted from the Dataframe
+    """
+    unit = units.loc[(wildcards.sample, wildcards.unit, wildcards.run, wildcards.lane)].dropna()
+    return unit
+
+
+def get_fastq_file(units: pd.DataFrame, wildcards: snakemake.io.Wildcards, read_pair: str = "fastq1") -> str:
```

Comment:
I try to never use global variables inside functions, one never know how the user modifies the variable. I was thinking that it could be a good option to move many of these functions to hydra-genetics/tools, since they most likely will be used in multiple modules/pipelines.

Reply:
Good point and I think it's better that way. We just need to agree on this style. However, if units is called units everywhere even in your top function it will still result in unexpected behavior if the user provides some wrong input and redefines `units`. But I still prefer to be more explicit. And including these functions in tools is probably also smart. So the library needs to be part of dependencies then, right?

Reply:
I will try to add hydra-genetics/tools to pip which will make it easy to add as a dependencies